python/sdist/amici/splines.py

```diff
@@ -0,0 +1,1687 @@
+            if length < 2:
+                raise ValueError("Length must be at least 2!")
+            step = (stop - start) / (length - 1)
+        elif length is not None:
+            raise ValueError("Only one of step/length can be specified!")
+        else:
+            step = sp.nsimplify(sp.sympify(step))
+
+        if start > stop:
+            raise ValueError(
+                f'Start point {start} greater than stop point {stop}!'
+            )
+
+        if step <= 0:
+            raise ValueError(f'Step size {step} must be strictly positive!')
+
+        xx = []
+        for i in count():
+            x = start + i * step
+            if not include_stop and x > stop:
+                break
+            xx.append(x)
+            if include_stop and x >= stop:
+                break
+
+        self._xx = np.asarray(xx)
+
+    @property
+    def start(self) -> sp.Basic:
+        """First point."""
+        return self._xx[0]
+
+    @property
+    def stop(self) -> sp.Basic:
+        """Last point."""
+        return self._xx[-1]
+
+    @property
+    def step(self) -> sp.Basic:
+        """Distance between consecutive points."""
+        if len(self._xx) > 1:
+            return self._xx[1] - self._xx[0]
+        return sp.core.numbers.Zero()
+
+    def __getitem__(self, i: Integral) -> sp.Basic:
+        return self._xx[i]
+
+    def __len__(self) -> int:
+        return len(self._xx)
+
+    def __array__(self, dtype=None) -> np.ndarray:
+        if dtype is None:
+            return self._xx
+        return np.array(self._xx, dtype=dtype)
+
+    def __repr__(self) -> str:
+        return (f"UniformGrid(start={self.start}, stop={self.stop}, "
+                f"step={self.step})")
+
+
+###############################################################################
+
+
+class AbstractSpline(ABC):
+    """
+    Base class for spline functions which can be computed efficiently
+    thanks to tailored C++ implementations in AMICI.
+    Inside an SBML file, such splines are implemented with
+    an assignment rule containing both a symbolic piecewise formula
+    for the spline (allowing compatibility with any SBML-aware software)
+    and annotations which encode the necessary information for AMICI to
+    recreate the spline object (allowing for fast computations when the SBML
+    file is used together with AMICI).
+    """
+
+    def __init__(
+            self,
+            sbml_id: Union[str, sp.Symbol],
+            x: Union[str, sp.Basic],
+            xx: Sequence,
+            yy: Sequence,
+            *,
+            bc: BClike = None,
+            extrapolate: BClike = None,
+            logarithmic_parametrization: bool = False
+    ):
+        """Base constructor for ``AbstractSpline`` objects.
+
+        :param sbml_id:
+            The SBML ID of the parameter associated to the spline
+            as a string or a SymPy symbol.
+
+        :param x:
+            The point at which the spline is evaluated.
```

Comment:
It was never tested for, but you are right that it should have been done. But I had a look at the C++ code and it seems that Paul hardcoded time as the point where the spline is evaluated, so the functionality from the Python interface (`splines.py`) is not carried over to `de_export.py` and C++. I guess Paul decided to concentrate on the low hanging fruit to make KoCo19 work before he left the group. I now raise errors when such splines are read (but still allow them to be created since the piecewise formulas should still be correct, with a warning that AMICI will not work on the resulting SBML files).

Reply:
:+1: